pkg/preemptiontoleration/preemption_toleration.go

```diff
@@ -0,0 +1,419 @@
+	defer func() {
+		metrics.PreemptionAttempts.Inc()
+		metrics.DeprecatedSchedulingAlgorithmPreemptionEvaluationDuration.Observe(metrics.SinceInSeconds(preemptionStartTime))
+	}()
+
+	nnn, err := pl.preempt(ctx, state, pod, m)
+	if err != nil {
+		return nil, framework.AsStatus(err)
+	}
+	if nnn == "" {
+		return nil, framework.NewStatus(framework.Unschedulable)
+	}
+	return &framework.PostFilterResult{NominatedNodeName: nnn}, framework.NewStatus(framework.Success)
+}
+
+// preempt finds nodes with pods that can be preempted to make room for "preemptor" to schedule.
+// This is almost identical to DefaultPreemption plugin's one.
+func (pl *PreemptionToleration) preempt(ctx context.Context, state *framework.CycleState, preemptor *v1.Pod, m framework.NodeToStatusMap) (string, error) {
+	cs := pl.fh.ClientSet()
+	ph := pl.fh.PreemptHandle()
+	nodeLister := pl.fh.SnapshotSharedLister().NodeInfos()
+
+	preemptor, err := pl.podLister.Pods(preemptor.Namespace).Get(preemptor.Name)
+	if err != nil {
+		klog.Errorf("Error getting the updated preemptor pod object: %v", err)
+		return "", err
+	}
+
+	// 1) Ensure the preemptor is eligible to preempt other pods.
+	if !dp.PodEligibleToPreemptOthers(preemptor, nodeLister, m[preemptor.Status.NominatedNodeName]) {
+		klog.V(5).Infof("Pod %v/%v is not eligible for more preemption.", preemptor.Namespace, preemptor.Name)
+		return "", nil
+	}
+
+	// 2) Find all preemption candidates.
+	candidates, err := pl.FindCandidates(ctx, state, preemptor, m)
+	if err != nil || len(candidates) == 0 {
+		return "", err
+	}
+
+	// 3) Interact with registered Extenders to filter out some candidates if needed.
+	candidates, err = dp.CallExtenders(ph.Extenders(), preemptor, nodeLister, candidates)
+	if err != nil {
+		return "", err
+	}
+
+	// 4) Find the best candidate.
+	bestCandidate := dp.SelectCandidate(candidates)
+	if bestCandidate == nil || len(bestCandidate.Name()) == 0 {
+		return "", nil
+	}
+
+	// 5) Perform preparation work before nominating the selected candidate.
+	if err := dp.PrepareCandidate(bestCandidate, pl.fh, cs, preemptor); err != nil {
+		return "", err
+	}
+
+	return bestCandidate.Name(), nil
+}
+
+// CanToleratePreemption evaluates whether the victimCandidate
+// pod can tolerate from preemption by the preemptor pod or not
+// by inspecting PriorityClass of victimCandidate pod.
+func CanToleratePreemption(
+	victimCandidate, preemptor *v1.Pod,
+	pcLister schedulinglisters.PriorityClassLister,
+	now time.Time,
+) (bool, error) {
+	preemptorPriority := corev1helpers.PodPriority(preemptor)
+	victimPriority := corev1helpers.PodPriority(victimCandidate)
+
+	if victimCandidate.Spec.PriorityClassName == "" {
+		return false, nil
+	}
+	victimPriorityClass, err := pcLister.Get(victimCandidate.Spec.PriorityClassName)
+	if err != nil {
+		return false, err
+	}
+
+	policy, err := GetCompletedPreemptionToleration(*victimPriorityClass)
+	if err != nil {
+		return false, nil
+	}
+	if policy == nil {
```

Comment:
We should respect PreemptionPolicy of preemptor even if victimPrioirtyClass doesn't have PreemptionToleration.
So, the case when this block return true, this function should return true even if victimPrioirtyClass doesn't have PreemptionToleration.

Reply:
Preemptor's `PreemptionPolicy` is [verified by `dp.PodEligibleToPreemptOthers()`](https://github.com/everpeace/scheduler-plugins/pull/2/files#diff-4c60b749a3f66515ec95911b45522ae0450f6c2e16dc8417256820cc1ebeda70R136), so we can assume that preemptor has nil or `PreemptLowerPriority` policy.

Reply:
Thank you!
https://github.com/kubernetes/kubernetes/blob/0d9c29078b71fafe65261cb8597e6da0b1c8fe9e/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

Sorry for my wrong comment.